### ISSUE TITLE
off-heap memory used by oakMap can not be released

### ISSUE THREAD

**Issue by fujian-zfj** [yahoo/Oak#192]
I use oakMap instead of concurrentSkipListMap in my project, while I find something strange.
When I no longer use an oakMap, java heap memory used by this oakMap is released, but off-heap memory does not seems to be released.

I use NMT to track the use of off-heap memory, and I find something strange. 
![image](https://user-images.githubusercontent.com/10379042/147069686-637a4ae7-6e1f-4ad7-bfb9-02c3622c131a.png)

![image](https://user-images.githubusercontent.com/10379042/147069535-c2f4022e-91a6-43db-9fef-c98140d40fa4.png)
ggggggggg

The heap memory occupied by current process is very small, but through the linux system command "top" found that it occupies huge memory. And I analyze the memory distribution in heap and find that heap memory used by the current process is very small. By the way, therer is no OakMap object in heap memory.
![image](https://user-images.githubusercontent.com/10379042/147070312-c8a2c583-159b-43d7-8d65-48ee3edb59d9.png)

**Comment by fujian-zfj** [comment]
private final OakMapBuilder<Long, AckData> oakMapBuilder;
private final OakMap<Long, AckData> unAckMessagesMap;

public OffHeapMessagePool(AmqpChannel amqpChannel) {
    this.amqpChannel = amqpChannel;
    this.oakMapBuilder = new OakMapBuilder<>(new OakLongComparator(), new OakLongSerializer(),
        new OakAckDataSerializer(), Long.MIN_VALUE).setMemoryCapacity(MEBIBYTE);
    this.unAckGetMessageSet = new ConcurrentSet<>();
    this.unAckMessagesMap = oakMapBuilder.buildOrderedMap();
}

public boolean addToUnAckMessageMap(long deliveryTag, AckData ackData) {
  return unAckMessagesMap.putIfAbsent(deliveryTag, ackData) == null ? true : false;
}

public boolean removeFromUnAckMessagesMap(long deliveryTag) {
  return unAckMessagesMap.remove(deliveryTag) != null ? true : false;
}

    public void clean() {
        if (unAckMessagesMap != null ) {
            try {
                unAckMessagesMap.close();
                //System.gc();
            } catch (Exception e){

            }
        }

    }

**Comment by liran-funaro** [comment]
Oak has a cached memory pool. It reuses the memory for other Oak instances.
After closing Oak the memory is not released back to the OS, but new Oak instances will reuse this memory.
Can you tell us more about your use case? Do you have a case where you don't create new Oak instances but need to release the memory back to the OS?

**Comment by fujian-zfj** [comment]
I create 100 OakMap. One thread keeps calling oakMap.putIfAbsent(key, value), and the other thread keeps calling oakMap.get(key) and aokMap.remove(key). During this period, no new  OakMap will be created.  A few days later, I found that my process is down due to outOfMemroy

**Comment by fujian-zfj** [comment]
![image](https://user-images.githubusercontent.com/10379042/147085568-76b1eb8f-8eac-469f-831c-857ed5d4a10b.png)

**Comment by fujian-zfj** [comment]
> 

Yes, I will not create new Oak Instances and need to release the memory back to the OS

**Comment by fujian-zfj** [comment]
I use MAT(Memory Analyzer Tool）to anlayze the memory usage of my process, and I find this:
![image](https://user-images.githubusercontent.com/10379042/147097172-b39a5a0e-d62e-40c4-8fcd-e8150248a2aa.png)
![image](https://user-images.githubusercontent.com/10379042/147097201-a7a44757-a685-4f2e-80c0-038353e1a2e2.png)

**Comment by sanastas** [comment]
Hi @fujian-zfj ,

Thanks for your interest in Oak and sorry for late reply. The issue that you see is because currently keys in the OakMap are not released and with the time going keys are consuming the memory. It was written that way because until now OakMap was used for relatively short periods of time and keys are usually not big.

However, we already have PR#188 where we are adding the new memory manager and the ability to release keys to OakMap. So it is mater of a week or so. If you need it really urgently, may be you would like to try OakHash (not ordered) instead of OakMap (ordered). In OakHash keys are released in time of the mapping remove. Are you using public release or current master?

**Comment by fujian-zfj** [comment]
I use current master. And what I need is OakMap(ConcurrentSkipListMap) instead of OakHash. 

I am looking forward to the new  memory manager which has the ability to release keys to OakMap.

If possible, I hope it can be launched as soon as possible, and I will continue to follow up on this issue.
